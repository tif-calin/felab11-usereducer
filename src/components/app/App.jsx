import React, { useReducer } from 'react';

const initialState = {
  before: [],
  current: '#ffffff',
  after: []
};

const reducer = (state, action) => {

  const { before, current, after } = state;

  switch (action.type) {
    case 'undo':
      return {
        before: [...before.slice(0, -1)],
        current: before[before.length - 1],
        after: [...after]
      };
    case 'redo':
      return {
        before: [...before],
        current: after[0],
        after: [...after.slice(1)]
      };
    case 'record':
      return {
        before: [...before, current],
        current: action.color,
        after: [] // not original functionality but i think it makes sense
      };
    default:
      return state;
  }
};


function App() {
  const [state, dispatch] = useReducer(reducer, initialState);

  const { current } = state;

  return <>
    <button onClick={() => dispatch({ type: 'undo' })}>undo</button>
    <button onClick={() => dispatch({ type: 'redo' })}>redo</button>
    <input
      type="color"
      value={current}
      onChange={({ target }) => 
        dispatch({ type: 'record', color: target.value })
      }
    />
    <div
      style={{ backgroundColor: current, width: '10rem', height: '10rem' }}
    >{current}</div>
  </>;
}

export default App;
