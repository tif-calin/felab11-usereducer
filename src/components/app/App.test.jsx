import React from 'react';
import { render, cleanup, fireEvent } from '@testing-library/react';
import App from './App';

describe('App component', () => {

  afterEach(() => cleanup());

  test('snapshot test for App', () => {
    const { asFragment } = render(<App />);
    expect(asFragment()).toMatchSnapshot();
  });

  test('can record, redo, and undo color changes', () => {
    const app = render(<App />);

    const undoButton = app.getByText('undo');
    const redoButton = app.getByText('redo');
    const displayDiv = app.getByText('#ffffff');
    const colorInput = app.getByDisplayValue('#ffffff');

    // change the color thrice
    fireEvent.change(colorInput, { target: { value: '#AABBCC' } });
    expect(displayDiv).toHaveStyle({ backgroundColor: '#AABBCC' });
    fireEvent.change(colorInput, { target: { value: '#11AA55' } });
    expect(displayDiv).toHaveStyle({ backgroundColor: '#11AA55' });
    fireEvent.change(colorInput, { target: { value: '#ABC123' } });
    expect(displayDiv).toHaveStyle({ backgroundColor: '#ABC123' });

    // click undo button twice
    fireEvent.click(undoButton);
    expect(displayDiv).toHaveStyle({ backgroundColor: '#11AA55' });
    fireEvent.click(undoButton);
    expect(displayDiv).toHaveStyle({ backgroundColor: '#AABBCC' });

    // click redo button
    fireEvent.click(redoButton);
    expect(displayDiv).toHaveStyle({ backgroundColor: '#11AA55' });
  });

});
