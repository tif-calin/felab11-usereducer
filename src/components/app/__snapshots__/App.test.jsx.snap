// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component snapshot test for App 1`] = `
<DocumentFragment>
  <button>
    undo
  </button>
  <button>
    redo
  </button>
  <input
    name="color"
    type="color"
    value="#ffffff"
  />
  <div
    style="background-color: rgb(255, 255, 255); width: 10rem; height: 10rem;"
  >
    #ffffff
  </div>
</DocumentFragment>
`;
